{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Assignment 1 Chapter 2.ipynb",
      "provenance": [],
      "collapsed_sections": [
        "H71heObrRXlO",
        "WEj-LvMPR7il",
        "Bkp10jJWR73S",
        "Cpoi50_wR8F9",
        "_n-RAldzR8Ti",
        "AxOMHvvDR8jD",
        "P5cl66_bR8wz",
        "nYJizVXqR8-k"
      ],
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "H71heObrRXlO"
      },
      "source": [
        "#### BA2A: Implement MotifEnumeration"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "lmmzKgfyQLaW",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "68db3d3b-09a4-437c-937b-2fbdf52fa964"
      },
      "source": [
        "def generateAllKmers(k):\n",
        "    base_neucleotides = [\"A\", \"T\", \"G\", \"C\"]\n",
        "    last_base = base_neucleotides #copying base neucleotides in last_base\n",
        "    current_array = [] #declare an emty list\n",
        "    for i in range(k-1): #for 2 length kmer the loop will run for i=0 and i=1\n",
        "        for j in base_neucleotides: #A, T, G, C\n",
        "            for k in last_base: #copy of A, T, G, C\n",
        "                current_array.append(j+k) \n",
        "        last_base = current_array #update last_base\n",
        "        current_array = [] #empty current array\n",
        "    return last_base\n",
        "\n",
        "def MismatchedMotifSet(text, k, d):\n",
        "    motifset = set()\n",
        "    all_Kmer = generateAllKmers(k) #generate all possible k mers\n",
        "    text_Kmer = textKmer(text, k)  #generate all possible kmers from given text\n",
        "    for i in range(len(all_Kmer)):\n",
        "        for j in range(len(text_Kmer)):\n",
        "          if hammingDistance(all_Kmer[i], text_Kmer[j]) <= d: #for each kmer generated from all possible kmers check hamming distance with main text kmers\n",
        "            motifset.add(all_Kmer[i]) \n",
        "    return motifset\n",
        "    \n",
        "def hammingDistance(text1, text2):\n",
        "    count=0;\n",
        "    for i in range(len(text1)):\n",
        "        if(text1[i] != text2[i]):\n",
        "            count+=1\n",
        "    return count\n",
        "\n",
        "def textKmer(text, k):\n",
        "    my_arr = []\n",
        "    for i in range(len(text)-k+1):\n",
        "        my_arr.append(text[i:i+k])\n",
        "    return my_arr\n",
        "\n",
        "def DNeighbourhood(genomelist, k,d):\n",
        "  count = {}\n",
        "\n",
        "  for genome in genomelist:\n",
        "    motifset = MismatchedMotifSet(genome, k, d)\n",
        "    for motif in motifset:\n",
        "      if motif not in count:\n",
        "        count[motif] = 1\n",
        "      else:\n",
        "        count[motif] += 1\n",
        "  \n",
        "  motiflist = []\n",
        "  for key in count:\n",
        "    if count[key] == len(genomelist):\n",
        "      motiflist.append(key)\n",
        "  return sorted(motiflist)\n",
        "\n",
        "def WriteToFile(result):\n",
        "  # outF = open(\"myOutFile.txt\", \"w\")\n",
        "  for line in result:\n",
        "    # write line to output file\n",
        "    print(line, end=\" \")\n",
        "\n",
        "k = 3\n",
        "d = 1\n",
        "with open('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Assignment/Chapter 2/rosalind_ba2a.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "length = len(genomelist)\n",
        "WriteToFile(DNeighbourhood(genomelist, k, d))\n"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "ATA ATT GTT TTT "
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "WEj-LvMPR7il"
      },
      "source": [
        "#### BA2B: Find a Median String"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MLBIRhtmR7rr",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b6f83fe8-bb52-4f5e-b8c0-c3a9b3f1a026"
      },
      "source": [
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] #remainder is the single symbol\n",
        "    pref_pattern = NumberToPattern(pref_index, k-1) #recursive call\n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def HammingDistance(text1, text2):\n",
        "    count=0;\n",
        "    for i in range(len(text1)):\n",
        "        if(text1[i] != text2[i]):\n",
        "            count+=1\n",
        "    return count\n",
        "\n",
        "def DistanceBetweenPatternsandGenome(genomelist, pattern):\n",
        "  k = len(pattern)\n",
        "  distance = 0\n",
        "  for genome in genomelist:\n",
        "    hammingDistance = 9999\n",
        "    genome_kmers = generateTextKmer(genome, k)\n",
        "    for kmer in  genome_kmers:\n",
        "      if hammingDistance > HammingDistance(pattern, kmer):\n",
        "        hammingDistance =  HammingDistance(pattern, kmer)\n",
        "    distance = distance+hammingDistance\n",
        "  #print(distance)  \n",
        "  return distance\n",
        "\n",
        "def MedianString(genomelist, k):\n",
        "  distance = 9999;\n",
        "  limit = pow(4,k)\n",
        "  for i in range(limit-1):\n",
        "    pattern = NumberToPattern(i,k)\n",
        "    if distance >= DistanceBetweenPatternsandGenome(genomelist, pattern):\n",
        "      #print(\"pattern, Distance\")\n",
        "      #print(pattern, distance)\n",
        "      distance = DistanceBetweenPatternsandGenome(genomelist, pattern)\n",
        "      median = pattern\n",
        "  return median\n",
        "\n",
        "k = int(input(\"Enter the length of kmer: \"))\n",
        "genomelist = []\n",
        "with open('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Assignment/Chapter 2/rosalind_ba2b.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "#print(genomelist)\n",
        "ans = MedianString(genomelist, k)\n",
        "print(ans)"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Enter the length of kmer: 3\n",
            "GAC\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Bkp10jJWR73S"
      },
      "source": [
        "#### BA2C: Find a Profile-most Probable k-mer in a String"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "AhgMF5cdR7_b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "2202abe6-2492-46ba-e252-7964d83f1e89"
      },
      "source": [
        "import numpy as np\n",
        "\n",
        "Genome = \"ACCTGTTTATTGCCTAAGTTCCGAACAAACCCAATATAGCCCGAGGGCCT\"\n",
        "k = 5\n",
        "profile = []\n",
        "print(\"Enter profile matrix value rowwise: \")\n",
        "\n",
        "arr_entry1 = list(map(float, input().strip().split()))\n",
        "profile =  np.array(arr_entry1).reshape(4,k)\n",
        "\n",
        "\n",
        "genomeList = list(Genome) # converting genome string into list\n",
        "\n",
        "\n",
        "## First plan is to generate all possible kmer  \n",
        "##\n",
        "##\n",
        "kMers = []\n",
        "for j in range(len(Genome)-k+1):\n",
        "    currentkmar = Genome[j:j+k] # make a substring\n",
        "    kMers.append(currentkmar)\n",
        "\n",
        "kMersSet = list(set(kMers))\n",
        "#print(kMers)\n",
        "\n",
        "porbableDictionary = {}\n",
        "for kMer in kMersSet:\n",
        "    probability = 1\n",
        "    kMerList = list(kMer)\n",
        "    for i in range(len(kMerList)):\n",
        "        if kMer[i] == 'A':\n",
        "            probability *= profile[0][i]\n",
        "        elif kMer[i] == 'C':\n",
        "            probability *= profile[1][i]\n",
        "        elif kMer[i] == 'G':\n",
        "            probability *= profile[2][i]\n",
        "        elif kMer[i] == 'T':\n",
        "            probability *= profile[3][i]\n",
        "    porbableDictionary[kMer] = probability\n",
        "indexOFmaxValue = max(porbableDictionary, key=porbableDictionary.get)\n",
        "print(\"The profile most probable\",k, \"-mer is:\", indexOFmaxValue)"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Enter profile matrix value rowwise: \n",
            "0.2 0.2 0.3 0.2 0.3 0.4 0.3 0.1 0.5 0.1 0.3 0.3 0.5 0.2 0.4 0.1 0.2 0.1 0.1 0.2\n",
            "The profile most probable 5 -mer is: CCGAG\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Cpoi50_wR8F9"
      },
      "source": [
        "#### BA2D: Implement GreedyMotifSearch"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Bk5u0oSuR8MG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ef07fc09-b2d9-48a8-a334-d1fd91a14222"
      },
      "source": [
        "import numpy as np\n",
        "\n",
        "\n",
        "def FormBestMotifs(genomes, k,t):\n",
        "  bmotifs = []\n",
        "  for genome in genomes:\n",
        "    bmotifs.append(genome[:k])\n",
        "  return bmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def FormCountMatrix(motifs): # count matrix is 4*k AAGG\n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    #print(motif)\n",
        "    for i in range(len(motif)):\n",
        "      # print(\"i: \", i)\n",
        "      # print(motif[i])\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def FormProfile(motifs): # profile matrix 4*k\n",
        "  # first we need to calculate count matrix\n",
        "  countMat = FormCountMatrix(motifs)\n",
        "  #print(\"countMat: \", countMat)\n",
        "  \n",
        "  profileMatrx = np.zeros((4,len(motifs[0])))\n",
        "  # now just divide count matrix by number of motifs\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrx[row][col] = countMat[row][col] / len(motifs)\n",
        "  #print(\"profile Matrix: \", profileMatrx)\n",
        "  return profileMatrx\n",
        "\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  #column wise\n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  #sotring the position\n",
        "        max = profile[row][col]  #finding maximum vale from column\n",
        "    consensus += NumberToSymbol(loc)  #creating string by concatanating maz count character\n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] #remainder is the single symbol\n",
        "    pref_pattern = numberToPattern(pref_index, k-1) #recursive call\n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = FormProfile(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  #motif from motifset\n",
        "    for i in range(len(motif)): #motif's ith charachter\n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def GreedyMotifSearch(genomes, k, t):\n",
        "  bestMotifs = FormBestMotifs(genomes, k,t)\n",
        "  #print(bestMotifs)\n",
        "  kmers = generateTextKmer(genomes[0], k)\n",
        "  #print(\"list of kmers in 1st: \", kmers)\n",
        "  for kmer in kmers:\n",
        "    motifs = []\n",
        "    motifs.append(kmer)\n",
        "    #print(\"kmer appended: \", kmer)\n",
        "    for idx in range(1, t):\n",
        "      # first form profile matrix with existing motifs\n",
        "      profile_matrix = FormProfile(motifs) \n",
        "      # calculate probable kmer and append that kmer to construct our motifs set\n",
        "      motifs.append(ProfileMostProbableKmer(genomes[idx], k, profile_matrix))\n",
        "    \n",
        "    # comparing current motif set with best motif set\n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "  return bestMotifs\n",
        "\n",
        "k = 3\n",
        "t = 5\n",
        "with open('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Lab Task 3/rosalind_ba2d.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "  #print(genomelist)\n",
        "ans = GreedyMotifSearch(genomelist, k, t)\n",
        "for i in ans:\n",
        "  print(i)"
      ],
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "CAG\n",
            "CAG\n",
            "CAA\n",
            "CAA\n",
            "CAA\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_n-RAldzR8Ti"
      },
      "source": [
        "#### BA2E: Implement GreedyMotifSearch with Pseudocounts"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TplSRZtkR8b7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "e5b55bd5-2542-4bec-f8e9-f4bb10f6e794"
      },
      "source": [
        "import numpy as np\n",
        "\n",
        "\n",
        "def FormBestMotifs(genomes, k,t):\n",
        "  bmotifs = []\n",
        "  for genome in genomes:\n",
        "    bmotifs.append(genome[:k])\n",
        "  return bmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "\n",
        "def FormCountMatrix(motifs): # count matrix is 4*k AAGG\n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    #print(motif)\n",
        "    for i in range(len(motif)):\n",
        "      # print(\"i: \", i)\n",
        "      # print(motif[i])\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def FormProfile(motifs): # profile matrix 4*k\n",
        "  # first we need to calculate count matrix\n",
        "  countMat = FormCountMatrix(motifs)\n",
        "  #print(\"count Mat: \", countMat)\n",
        "  pseudoMat = PseudoCountMatrix(countMat)\n",
        "  profileMatrx = np.zeros((4,len(motifs[0])))\n",
        "  # now just divide count matrix by number of motifs\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrx[row][col] = pseudoMat[row][col] / len(motifs)\n",
        "  return profileMatrx\n",
        "\n",
        "def PseudoCountMatrix(countMat):\n",
        "  pseudo = [1*4]*len(countMat[0])\n",
        "  pseudoMatrix = pseudo+countMat\n",
        " #print(\"pseudo\", pseudo)\n",
        "  return pseudoMatrix\n",
        "\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  #column wise\n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  #sotring the position\n",
        "        max = profile[row][col]  #finding maximum vale from column\n",
        "    consensus += NumberToSymbol(loc)  #creating string by concatanating maz count character\n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] #remainder is the single symbol\n",
        "    pref_pattern = numberToPattern(pref_index, k-1) #recursive call\n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = FormProfile(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  #motif from motifset\n",
        "    for i in range(len(motif)): #motif's ith charachter\n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def GreedyMotifSearchWithPseudoCount(genomes, k, t):\n",
        "  bestMotifs = FormBestMotifs(genomes, k,t)\n",
        "  #print(bestMotifs)\n",
        "  kmers = generateTextKmer(genomes[0], k)\n",
        "  #print(\"list of kmers in 1st: \", kmers)\n",
        "  for kmer in kmers:\n",
        "    motifs = []\n",
        "    motifs.append(kmer)\n",
        "    #print(\"kmer appended: \", kmer)\n",
        "    for idx in range(1, t):\n",
        "      # first form profile matrix with existing motifs\n",
        "      profile_matrix = FormProfile(motifs) \n",
        "      # calculate probable kmer and append that kmer to construct our motifs set\n",
        "      motifs.append(ProfileMostProbableKmer(genomes[idx], k, profile_matrix))\n",
        "    \n",
        "    # comparing current motif set with best motif set\n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "  return bestMotifs\n",
        "\n",
        "def WriteToFile(result):\n",
        "  for line in result:\n",
        "    print(line)\n",
        "\n",
        "k = 3\n",
        "t = 5\n",
        "with open('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Assignment/Chapter 2/rosalind_ba2e.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "  #print(genomelist)\n",
        "ans = GreedyMotifSearchWithPseudoCount(genomelist, k, t)\n",
        "WriteToFile(ans)"
      ],
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TCA\n",
            "TCA\n",
            "TCG\n",
            "TCA\n",
            "TCG\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "AxOMHvvDR8jD"
      },
      "source": [
        "#### BA2F: Implement RandomizedMotifSearch"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MRT9m13wR8pt",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5e1481d2-0bef-4dca-c284-d0c97990d20c"
      },
      "source": [
        "import numpy as np\n",
        "from random import randint\n",
        "\n",
        "def FormBestMotifs(genomes, k,t):\n",
        "  bmotifs = []\n",
        "  for genome in genomes:\n",
        "    bmotifs.append(genome[:k])\n",
        "  return bmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "\n",
        "def FormCountMatrix(motifs): # count matrix is 4*k AAGG\n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    #print(motif)\n",
        "    for i in range(len(motif)):\n",
        "      # print(\"i: \", i)\n",
        "      # print(motif[i])\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def FormProfile(motifs): # profile matrix 4*k\n",
        "  # first we need to calculate count matrix\n",
        "  countMat = FormCountMatrix(motifs)\n",
        "  #print(\"count Mat: \", countMat)\n",
        "  pseudoMat = PseudoCountMatrix(countMat)\n",
        "  profileMatrx = np.zeros((4,len(motifs[0])))\n",
        "  # now just divide count matrix by number of motifs\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrx[row][col] = pseudoMat[row][col] / len(motifs)\n",
        "  return profileMatrx\n",
        "\n",
        "def PseudoCountMatrix(countMat):\n",
        "  pseudo = [1*4]*len(countMat[0])\n",
        "  pseudoMatrix = pseudo+countMat\n",
        " #print(\"pseudo\", pseudo)\n",
        "  return pseudoMatrix\n",
        "\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  #column wise\n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  #sotring the position\n",
        "        max = profile[row][col]  #finding maximum vale from column\n",
        "    consensus += NumberToSymbol(loc)  #creating string by concatanating maz count character\n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] #remainder is the single symbol\n",
        "    pref_pattern = numberToPattern(pref_index, k-1) #recursive call\n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = FormProfile(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  #motif from motifset\n",
        "    for i in range(len(motif)): #motif's ith charachter\n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def getRandom(n):\n",
        "  return randint(0, n)\n",
        "\n",
        "def FormRandomlyMotifs(genomes, k):\n",
        "  motifs = []\n",
        "  for genome in genomes:\n",
        "    start = getRandom(len(genome)-k)\n",
        "    motifs.append(genome[start:start+k])\n",
        "  return motifs\n",
        "\n",
        "def generateMotifsSet(profile_matrix, k, genomes):\n",
        "  probKmers = []\n",
        "  for genome in genomes:\n",
        "    probKmers.append(ProfileMostProbableKmer(genome, k, profile_matrix))\n",
        "  return probKmers\n",
        "\n",
        "def RandomizedMotifSearch(genomes, k, t):\n",
        "  motifs = FormRandomlyMotifs(genomes, k)\n",
        "  bestMotifs = motifs\n",
        "  while True:\n",
        "    profile_matrix = FormProfile(motifs)\n",
        "    motifs = generateMotifsSet(profile_matrix, k, genomes)\n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "    else:\n",
        "      return bestMotifs\n",
        "\n",
        "def WriteToFile(result):\n",
        "  for line in result:\n",
        "    print(line)\n",
        "\n",
        "# main function\n",
        "k = 8\n",
        "t = 5\n",
        "with open('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Lab Task 4/rosalind_ba2f.txt', 'r') as f:\n",
        "  genomelist = [line.strip() for line in f.readlines()]\n",
        "  \n",
        "best_score = float(\"Inf\")\n",
        "best_result = []\n",
        "for i in range(1000):\n",
        "  ans = RandomizedMotifSearch(genomelist, k, t)\n",
        "  curr_score = Score(ans)\n",
        "  if curr_score <= best_score:\n",
        "    best_score = curr_score\n",
        "    best_result = ans\n",
        "WriteToFile(best_result)"
      ],
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TCTCGGGG\n",
            "CCAAGGTG\n",
            "TACAGGCG\n",
            "TTCAGGTG\n",
            "TCCACGTG\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P5cl66_bR8wz"
      },
      "source": [
        "#### BA2G: Implement GibbsSampler"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "A7fQZP_KR83E",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9af9841d-05a7-46de-cab9-52e4cfb921d0"
      },
      "source": [
        "import numpy as np\n",
        "import random\n",
        "\n",
        "def FormBestMotifs(genomes, k,t):\n",
        "  bmotifs = []\n",
        "  for genome in genomes:\n",
        "    bmotifs.append(genome[:k])\n",
        "  return bmotifs\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "\n",
        "def FormCountMatrix(motifs): # count matrix is 4*k AAGG\n",
        "  count = np.zeros((4,len(motifs[0])))\n",
        "  for motif in motifs:\n",
        "    #print(motif)\n",
        "    for i in range(len(motif)):\n",
        "      # print(\"i: \", i)\n",
        "      # print(motif[i])\n",
        "      if motif[i-1] == 'A':\n",
        "        count[0][i-1] += 1\n",
        "      elif motif[i-1] == 'C':\n",
        "        count[1][i-1] += 1\n",
        "      elif motif[i-1] == 'G':\n",
        "        count[2][i-1] += 1\n",
        "      elif motif[i-1] == 'T':\n",
        "        count[3][i-1] += 1\n",
        "  return count\n",
        "\n",
        "def FormProfile(motifs): # profile matrix 4*k\n",
        "  # first we need to calculate count matrix\n",
        "  countMat = FormCountMatrix(motifs)\n",
        "  #print(\"count Mat: \", countMat)\n",
        "  pseudoMat = PseudoCountMatrix(countMat)\n",
        "  profileMatrx = np.zeros((4,len(motifs[0])))\n",
        "  # now just divide count matrix by number of motifs\n",
        "  for row in range(4):\n",
        "    for col in range(k):\n",
        "      profileMatrx[row][col] = pseudoMat[row][col] / len(motifs)\n",
        "  return profileMatrx\n",
        "\n",
        "def PseudoCountMatrix(countMat):\n",
        "  pseudo = [1*4]*len(countMat[0])\n",
        "  pseudoMatrix = pseudo+countMat\n",
        " #print(\"pseudo\", pseudo)\n",
        "  return pseudoMatrix\n",
        "\n",
        "\n",
        "def ProfileMostProbableKmer(genome, k, profile_matrix):\n",
        "  maxprob = 0\n",
        "  kmer = genome[0:k]\n",
        "  for i in range(1,len(genome) - k +1):\n",
        "    prob = 1\n",
        "    pattern = genome[i:i+k]\n",
        "    for j in range(k):\n",
        "      l = symbolToNumber(pattern[j])\n",
        "      prob *= profile_matrix[l][j]\n",
        "    if prob > maxprob:\n",
        "      maxprob = prob\n",
        "      kmer = pattern\n",
        "  return kmer\n",
        "  \n",
        "def symbolToNumber(symbol):\n",
        "\tif symbol == \"A\":\n",
        "\t\treturn 0\n",
        "\tif symbol == \"C\":\n",
        "\t\treturn 1\n",
        "\tif symbol == \"G\":\n",
        "\t\treturn 2\n",
        "\tif symbol == \"T\":\n",
        "\t\treturn 3\n",
        "\n",
        "def NumberToSymbol(number):\n",
        "\tif number == 0:\n",
        "\t\treturn \"A\"\n",
        "\tif number == 1:\n",
        "\t\treturn \"C\"\n",
        "\tif number == 2:\n",
        "\t\treturn \"G\"\n",
        "\tif number == 3:\n",
        "\t\treturn \"T\"\n",
        "\n",
        "def Consensus(profile):\n",
        "  consensus = \"\"\n",
        "  for col in range(len(profile[0])):  #column wise\n",
        "    max = 0\n",
        "    loc = 0\n",
        "    for row in range(4):\n",
        "      if profile[row][col] > max:\n",
        "        loc = row  #sotring the position\n",
        "        max = profile[row][col]  #finding maximum vale from column\n",
        "    consensus += NumberToSymbol(loc)  #creating string by concatanating maz count character\n",
        "  return consensus\n",
        "\n",
        "neucleotide_base = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
        "def NumberToPattern(index, k):\n",
        "    if(k == 1): #base case\n",
        "        return neucleotide_base[index]\n",
        "    pref_index = int(index/4)\n",
        "    rem = int(index%4)\n",
        "    symbol = neucleotide_base[rem] #remainder is the single symbol\n",
        "    pref_pattern = numberToPattern(pref_index, k-1) #recursive call\n",
        "    \n",
        "    return pref_pattern+symbol\n",
        "\n",
        "def Score(motifs):\n",
        "  profile = FormProfile(motifs)\n",
        "  cons = Consensus(profile)\n",
        "  score = 0\n",
        "  for motif in motifs:  #motif from motifset\n",
        "    for i in range(len(motif)): #motif's ith charachter\n",
        "      if cons[i] != motif[i]:\n",
        "        score+=1\n",
        "  return score\n",
        "\n",
        "def getRandom(n):\n",
        "  return randint(0, n)\n",
        "\n",
        "def FormRandomlyMotifs(genomes, k):\n",
        "  motifs = []\n",
        "  for genome in genomes:\n",
        "    start = getRandom(len(genome)-k)\n",
        "    motifs.append(genome[start:start+k])\n",
        "  return motifs\n",
        "\n",
        "def generateMotifsSet(profile_matrix, k, genomes):\n",
        "  probKmers = []\n",
        "  for genome in genomes:\n",
        "    probKmers.append(ProfileMostProbableKmer(genome, k, profile_matrix))\n",
        "  return probKmers\n",
        "\n",
        "def ProfileRandomGeneratedKmer(profile_matrix, k, genome):\n",
        "  kmers = generateTextKmer(genome, k)\n",
        "  probs = []\n",
        "  for kmer in kmers:\n",
        "    probs.append(Probability(profile_matrix, kmer))\n",
        "  kmer = WeightedRandom(probs, genome,k)\n",
        "  return kmer\n",
        "\n",
        "def WeightedRandom(probs, genome, k):\n",
        "  s = sum(probs)\n",
        "  r = random.random()\n",
        "  counter = 0.0\n",
        "  for i in range(len(probs)):\n",
        "    counter += probs[i]\n",
        "    if counter/s >= r:\n",
        "      return genome[i:i+k]\n",
        "\n",
        "def Probability(profile_matrix, kmer):\n",
        "  prob = 1.0\n",
        "  for i in range(len(kmer)):\n",
        "    x = symbolToNumber(kmer[i])\n",
        "    prob = prob*profile_matrix[x][i]\n",
        "  return prob\n",
        "\n",
        "def GibbsSampler(genomes, k, t, N):\n",
        "  motifs = FormRandomlyMotifs(genomes, k)\n",
        "  bestMotifs = motifs\n",
        "  for j in range(N):\n",
        "    i = getRandom(t-1)\n",
        "    profile_matrix = FormProfile(motifs[0:i] + motifs[i+1:])\n",
        "    motifi = ProfileRandomGeneratedKmer(profile_matrix, k, genomes[i])\n",
        "    motifs[i] = motifi\n",
        "    if Score(motifs) < Score(bestMotifs):\n",
        "      bestMotifs = motifs\n",
        "    else:\n",
        "      return bestMotifs\n",
        "\n",
        "def WriteToFile(result):\n",
        "  for line in result:\n",
        "    print(line)\n",
        "    \n",
        "# main function\n",
        "k = 8        \n",
        "t = 5        \n",
        "N = 100      \n",
        "with open ('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Lab Task 4/rosalind_ba2g.txt', 'r') as f:\n",
        "    inputLines = [line.strip() for line in f.readlines()]\n",
        "  #print(genomelist)\n",
        "best = GibbsSampler(inputLines, k, t, N)\n",
        "best_score = Score(best)\n",
        "best_result = []\n",
        "for i in range(20):\n",
        "  ans = GibbsSampler(inputLines, k, t, N)\n",
        "  curr_score = Score(ans)\n",
        "  if curr_score <= best_score:\n",
        "    best_score = curr_score\n",
        "    best_result = ans\n",
        "WriteToFile(best_result)"
      ],
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TAAACGGC\n",
            "AAGGTGCC\n",
            "ACCGAGAC\n",
            "GTCGGTGA\n",
            "CACGTGCA\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nYJizVXqR8-k"
      },
      "source": [
        "#### BA2H: Implement DistanceBetweenPatternAndStrings"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ca6eKtD8R9FL",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f017b561-1c88-4ebf-a2f2-1962b7974467"
      },
      "source": [
        "def DistanceBetweenPatternsandGenome(genomelist, pattern):\n",
        "  k = len(pattern)\n",
        "  distance = 0\n",
        "  for genome in genomelist:\n",
        "    hammingDistance = float(\"inf\")\n",
        "    genome_kmers = generateTextKmer(genome, k)\n",
        "    for kmer in  genome_kmers:\n",
        "      if hammingDistance > HammingDistance(pattern, kmer):\n",
        "        hammingDistance =  HammingDistance(pattern, kmer)\n",
        "    distance = distance+hammingDistance\n",
        "  #print(distance)  \n",
        "  return distance\n",
        "\n",
        "def generateTextKmer(text, k):\n",
        "  kmers = set()\n",
        "  for i in range(len(text)-k+1):\n",
        "    kmers.add(text[i:i+k])\n",
        "  return kmers\n",
        "\n",
        "def HammingDistance(text1, text2):\n",
        "    count=0;\n",
        "    for i in range(len(text1)):\n",
        "        if(text1[i] != text2[i]):\n",
        "            count+=1\n",
        "    return count\n",
        "\n",
        "\n",
        "#pattern = input()\n",
        "with open('/content/drive/MyDrive/Bioinformatics Lab/Rosalind Dataset/Assignment/Chapter 2/rosalind_ba2h.txt', 'r') as f:\n",
        "  pattern = f.readline().strip()\n",
        "  listWords = f.readline().strip().split(\" \")\n",
        "  #print(listWords)\n",
        "\n",
        "\n",
        "print(DistanceBetweenPatternsandGenome(listWords, pattern))\n"
      ],
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "5\n"
          ],
          "name": "stdout"
        }
      ]
    }
  ]
}